daemon/internal/cloud_client/trace_event_request.go

```diff
@@ -0,0 +1,215 @@
+type TraceEventDetails struct {
+	// How many were parsed as valid trace events. Currently, we assume that a
+	// JSON object is a valid trace event if it has either a "request" or a
+	// "response" field.
+	Parsed int `json:"parsed"`
+
+	// How many were dropped because the queue was full.
+	Drops int `json:"drops"`
+}
+
+func (req traceEventRequest) handle(client *cloudClient) {
+	// See if the service and trace are known by the daemon. If the daemon
+	// doesn't know about the service or the trace yet, either the client is
+	// misbehaving or the daemon has restarted and lost its state. Either way,
+	// the daemon can't accept events until it learns about the trace's logging
+	// options from the cloud.
+	//
+	// If the daemon has been restarted, things will normalize when the first
+	// round of polling occurs, and the client can then resume sending trace
+	// events. Until that happens, just reject the trace events.
+	serviceInfo, traceInfo := client.getInfo(req.serviceID, req.traceID)
+	if serviceInfo == nil || traceInfo == nil {
+		defer close(req.responseChannel)
+		var message string
+		if serviceInfo == nil {
+			message = fmt.Sprintf("Service %q was not previously registered with this daemon", akid.String(req.serviceID))
+		} else {
+			message = fmt.Sprintf("Trace %q was not previously registered with this daemon", akid.String(req.traceID))
+		}
+
+		req.responseChannel <- newTraceEventResponse(http.StatusBadRequest, message, nil)
+		return
+	}
+
+	// Register the client with the trace.
+	traceInfo.clientNames[req.clientName] = struct{}{}
+
+	// Start a goroutine for relaying the incoming trace events to the collector.
+	traceEventChannel := traceInfo.traceEventChannel
+	go func() {
+		noMoreEvents := false
+		numTraceEventsParsed := 0
+		numTraceEventsDropped := 0
+
+		var err error = nil
+		for {
+			// Deserialize the next trace event.
+			var traceEvent TraceEvent
+			if err = req.traceEvents.Decode(&traceEvent); err != nil {
+				if err == io.EOF {
+					// We've reached the end of the stream, which isn't really an error.
+					// However, until we receive an empty JSON object from the client, we
+					// expect to receive more trace events from the client in a
+					// subsequent connection.
+					err = nil
+				}
+				break
+			}
+
+			// If the deserialized object has neither a request nor a response, treat
+			// it as signalling the end of the trace from this client.
+			if traceEvent.Request == nil && traceEvent.Response == nil {
+				noMoreEvents = true
+				break
+			}
+
+			numTraceEventsParsed++
+
+			// Attempt to enqueue the trace event.
+			select {
+			case traceEventChannel <- &traceEvent:
+			default:
+				numTraceEventsDropped++
+			}
+		}
+
+		// Log any errors that we encountered while processing the trace events.
+		eventDetails := TraceEventDetails{
+			Parsed: numTraceEventsParsed,
+			Drops:  numTraceEventsDropped,
+		}
+
+		// Send the result to the client.
+		{
+			defer close(req.responseChannel)
+
```

Comment:
I don't know if you are expecting the defer to exit at the end of the block, but it does not.  (If not, I don't understand why this section is in a curly-brace block.)

Reply:
I'm using the braces to demarcate the region of code to which the `Send the result to the client` comment applies.